sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala

```diff
@@ -225,12 +316,44 @@ private[parquet] class ParquetFilters(pushDownDate: Boolean, pushDownStartWith:
   def createFilter(schema: MessageType, predicate: sources.Filter): Option[FilterPredicate] = {
     val nameToType = getFieldMap(schema)
 
+    def isDecimalMatched(value: Any, decimalMeta: DecimalMetadata): Boolean = value match {
+      case decimal: JBigDecimal =>
+        decimal.scale == decimalMeta.getScale
+      case _ => false
+    }
+
+    // Since SPARK-24716, ParquetFilter accepts parquet file schema to convert to
+    // data source Filter. This must make sure that filter value matched the Filter.
+    // If doesn't matched, then the schema used to read the file is incorrect,
+    // which would cause data corruption.
+    def valueCanMakeFilterOn(name: String, value: Any): Boolean = {
+      value == null || (nameToType(name) match {
+        case ParquetBooleanType => value.isInstanceOf[JBoolean]
+        case ParquetByteType | ParquetShortType | ParquetIntegerType => value.isInstanceOf[Number]
+        case ParquetLongType => value.isInstanceOf[JLong]
+        case ParquetFloatType => value.isInstanceOf[JFloat]
+        case ParquetDoubleType => value.isInstanceOf[JDouble]
+        case ParquetStringType => value.isInstanceOf[String]
+        case ParquetBinaryType => value.isInstanceOf[Array[Byte]]
+        case ParquetDateType => value.isInstanceOf[Date]
+        case ParquetSchemaType(DECIMAL, INT32, 0, decimalMeta) =>
```

Comment:
Can the decimal cases be collapsed to a single case on `ParquetSchemaType(DECIMAL, _, _, decimalMetadata)`?

Reply:
No.
![image](https://user-images.githubusercontent.com/5399861/42616795-0c40b9e8-85e2-11e8-9697-ab5572b66bab.png)

Reply:
Have you tried not using `|` and ignoring the physical type with `_`?